lib/OpenQA/Utils.pm

```diff
@@ -278,65 +278,83 @@ sub log_fatal {
 }
 
+sub _current_log_level {
+    # FIXME: avoid use of $app here
+    return defined $app && $app->can('log') && $app->log->can('level') && $app->log->level;
+}
+
 # The %options parameter is used to control which destinations the message should go.
 # Accepted parameters: channels, standard.
 #  - channels. Scalar or a arrayref containing the name of the channels to log to.
-#  - standard. Boolean to indicate if it should use the *defaults*  to log.
+#  - standard. Boolean to indicate if it should use the *defaults* to log.
 #
-#  This function is used together with _log, and if any of parameters above don't exist,
-#  then it will log to the defaults (by default is $app) - the standard option need to be
-#  set to true. Please check the function add_log_channel to learn on how to set a
-#  channel as default
+# If any of parameters above don't exist, this function will log to the defaults (by
+# default it is $app). The standard option need to be set to true. Please check the function
+# add_log_channel to learn on how to set a channel as default.
 sub _log_msg {
     my ($level, $msg, %options) = @_;
-    my $log_to_standard = 0;
 
+    # use default options
     if (!%options) {
-        # set defaults
-        _log_msg($level, $msg, channels => $log_defaults{CHANNELS}, standard => $log_defaults{LOG_TO_STANDARD_CHANNEL});
-        $log_to_standard = 0;
+        return _log_msg(
+            $level, $msg,
+            channels => $log_defaults{CHANNELS},
+            standard => $log_defaults{LOG_TO_STANDARD_CHANNEL});
     }
-    else {
-        # TODO: We need to get rid of $app here :(
-        $msg = "[pid:$$] " . $msg
-          if defined $app && $app->can('log') && $app->log->can('level') && $app->log->level eq 'debug';
-        if ($options{channels}) {
-            if (ref($options{channels}) eq 'ARRAY') {
-                for my $channel (@{$options{channels}}) {
-                    _log($level, $msg, $channel);
-                }
-            }
-            else {
 
-                _log($level, $msg, $options{channels});
-            }
+    # prepend process ID on debug level
+    if (_current_log_level eq 'debug') {
+        $msg = "[pid:$$] $msg";
+    }
+
+    # log to channels
+    my $wrote_to_at_least_one_channel = 0;
+    if (my $channels = $options{channels}) {
+        for my $channel (ref($channels) eq 'ARRAY' ? @$channels : $channels) {
+            $wrote_to_at_least_one_channel |= _log_to_channel_by_name($level, $msg, $channel);
         }
-        $log_to_standard = $options{standard} // $log_defaults{LOG_TO_STANDARD_CHANNEL};
     }
 
-    _log($level, $msg) if $log_to_standard;
+    # log to standard (as fallback or when explicitely requested)
+    if (!$wrote_to_at_least_one_channel || ($options{standard} // $log_defaults{LOG_TO_STANDARD_CHANNEL})) {
+        # use Mojolicious app if available and otherwise just STDERR/STDOUT
+        _log_via_mojo_app($level, $msg) or _log_to_stderr_or_stdout($level, $msg);
```

Comment:
Why try to still go via mojo_app if it could not log to channels?

Reply:
Maybe the `Mojo::Log` object of the application would work (in contrast to the `Mojo::Log` objects of the desired channels)?

Besides, I didn't intent to change the fallback hierarchy itself here. I just make sure the fallbacks are also used in case of an exception.